File: apigateway/models/test_config.go
package models

import "time"

// TestConfig model
type TestConfig struct {
	ID    uint    `json:"id" gorm:"primaryKey"`
	Name  string  `json:"name"`
	Tests []*Test `json:"tests"`
}

// Test config to make requests
type Test struct {
	ID                   uint            `json:"id" gorm:"primaryKey"`
	TestConfigID         uint            `gorm:"foreignKey:TestConfigID"`
	URL                  string          `json:"url"`
	Method               string          `json:"method"`
	Payload              string          `json:"payload,omitempty"`
	RequestCount         int             `json:"requestCount"`
	GoroutineCount       int             `json:"goroutineCount"`
	ExpectedResponseCode uint            `json:"expectedResponseCode"`
	ExpectedResponseBody string          `json:"expectedResponseBody"`
	TransportConfig      TransportConfig `json:"transportConfig" gorm:"foreignKey:TestID"`
}

// RunTest tests
type RunTest struct {
	ID        uint `json:"id" gorm:"primaryKey"`
	TestID    uint `gorm:"foreignKey:TestID"`
	StartTime *time.Time
	EndTime   *time.Time
	Passed    bool `json:"passed"`
}

// TransportConfig is used to specify how to make requests
type TransportConfig struct {
	TestID            uint `json:"test_id"`
	DisableKeepAlives bool `json:"DisableKeepAlives"`
}

File: web/src/api/entity/test_config.ts
import { makeReq } from "../api";

export interface TestConfig {
  id?: number;
  name: string;
  tests: Test[];
}
export interface Test {
  id?: number;
  requestCount: number;
  goroutineCount: number;
  url: string;
  method: string;
  payload: string;
  expectedResponseCode: number;
  expectedResponseBody: string;
  transportConfig: TransportConfig;
}

export interface TransportConfig {
  DisableKeepAlives: boolean;
}

export const saveTests = async (testConfig: TestConfig) => {
  return await makeReq("/tests", "POST", testConfig);
};

export const updateTests = async (testConfig: TestConfig) => {
  return await makeReq("/tests", "PUT", testConfig);
};

export const runTests = async (testConfig: TestConfig) => {
  return await makeReq("/tests/start", "POST", testConfig);
};

export const listTests = async () => {
  return await makeReq("/tests", "Get");
};

export const deleteTestsReq = async (testConfig: TestConfig) => {
  return await makeReq("/tests", "DELETE", testConfig);
};

export const deleteTestReq = async (test: Test) => {
  return await makeReq("/tests/test", "DELETE", test);
};

export const updateTestReq = async (test: Test) => {
  return await makeReq("/tests/test", "PUT", test);
};
